### ISSUE TITLE
[Feature Request] Autodiff should check for `nan` values during backward run

### ISSUE THREAD

**Issue by jim19930609** [taichi-dev/taichi#7116]
## What's the issue

Modified from `billiards.py`, you'll observe `nan` in the gradient.
```
import math
import numpy as np
import os
import matplotlib.pyplot as plt 

real = ti.f32
ti.init(default_fp=real, offline_cache=False, flatten_if=True, print_ir=True)

scalar = lambda: ti.field(dtype=real)
vec = lambda: ti.Vector.field(2, dtype=real)

loss = scalar()
init_x = vec()
impulse = vec()

ti.root.place(init_x)
ti.root.place(loss)
ti.root.place(impulse)
ti.root.lazy_grad()

@ti.kernel
def collide():
    imp = ti.Vector([0.0, 0.0])
    dist = init_x[None]
    cond = dist[0] < 0.06
    if cond:
        dir = (1.0 / dist.sum()) * dist
        imp = 0.5 * dir 
        #print(1111)
    impulse[None] = imp 


init_x[None] = [0.1, 0.5]
with ti.ad.Tape(loss):
    collide()

print(init_x.grad[None][0])
print(init_x.grad[None][1])
```

## Diagnose
Suppose we have the following CHI IR:
```
$1 = alloca
$2 : local store $1, 0.0
if (cond) {
  $3 : local store $1, 10
  $4 = local load $1
  $5 = div 1.0, $4
  
  $6 = local store $1, $5
}
```

The `If clause` will be flattened by `irpass::simplify()`, resulting in:
```
$1 = alloca
$2 : local store $1, 0.0

$3 = local load $1
$4 = select cond, 0.0, $3
$5 = local store $1, $4

$6 = local load $1
$7 = div 1.0, $6

$8 = local load $1
$9 = select cond, $7, $8
$10 : local store $1, $9
```

The problem is that when `cond = false`, you'll get `$7 = 1.0 / 0.0` thus $7 will be `nan`. During backward run, statement $7 will be a dangling entrance with `grad($7) = 0`, thus `grad($6) = 0 * nan = nan`.

To avoid surprising the users, `autodiff` better check for `nan` values during backward run at least under `debug` mode.
